optaplanner-quarkus-integration/optaplanner-quarkus/deployment/src/main/java/org/optaplanner/quarkus/deployment/OptaPlannerProcessor.java

```diff
@@ -210,32 +213,95 @@ private void applySolverProperties(RecorderContext recorderContext,
         if (solverConfig.getScoreDirectorFactoryConfig() == null) {
-            ScoreDirectorFactoryConfig scoreDirectorFactoryConfig = new ScoreDirectorFactoryConfig();
-            scoreDirectorFactoryConfig.setEasyScoreCalculatorClass(
-                    findImplementingClass(DotNames.EASY_SCORE_CALCULATOR, indexView));
-            scoreDirectorFactoryConfig.setConstraintProviderClass(
-                    findImplementingClass(DotNames.CONSTRAINT_PROVIDER, indexView));
-            scoreDirectorFactoryConfig.setIncrementalScoreCalculatorClass(
-                    findImplementingClass(DotNames.INCREMENTAL_SCORE_CALCULATOR, indexView));
-            ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-            if (classLoader.getResource(SolverBuildTimeConfig.DEFAULT_SCORE_DRL_URL) != null) {
-                scoreDirectorFactoryConfig.setScoreDrlList(Collections.singletonList(
-                        SolverBuildTimeConfig.DEFAULT_SCORE_DRL_URL));
+            ScoreDirectorFactoryConfig scoreDirectorFactoryConfig =
+                    defaultScoreDirectoryFactoryConfig(indexView, effectiveConstraintsDrl);
+            solverConfig.setScoreDirectorFactoryConfig(scoreDirectorFactoryConfig);
+        } else {
+            ScoreDirectorFactoryConfig scoreDirectorFactoryConfig = solverConfig.getScoreDirectorFactoryConfig();
+            if (constraintsDrlFromProperty.isPresent()) {
+                scoreDirectorFactoryConfig.setScoreDrlList(Collections.singletonList(constraintsDrlFromProperty.get()));
+            } else {
+                if (scoreDirectorFactoryConfig.getScoreDrlList() == null) {
+                    defaultConstraintsDrl.ifPresent(resolvedConstraintsDrl -> scoreDirectorFactoryConfig
+                            .setScoreDrlList(Collections.singletonList(resolvedConstraintsDrl)));
+                }
             }
-            if (scoreDirectorFactoryConfig.getEasyScoreCalculatorClass() == null
-                    && scoreDirectorFactoryConfig.getConstraintProviderClass() == null
-                    && scoreDirectorFactoryConfig.getIncrementalScoreCalculatorClass() == null
-                    && scoreDirectorFactoryConfig.getScoreDrlList() == null) {
-                throw new IllegalStateException("No classes found that implement "
-                        + EasyScoreCalculator.class.getSimpleName() + ", "
-                        + ConstraintProvider.class.getSimpleName() + " or "
-                        + IncrementalScoreCalculator.class.getSimpleName() + ", nor a "
-                        + SolverBuildTimeConfig.DEFAULT_SCORE_DRL_URL + " resource.");
+        }
+
+        if (solverConfig.getScoreDirectorFactoryConfig().getScoreDrlList() != null) {
+            boolean isDroolsDynamicPresent = isClassDefined("org.drools.dynamic.DynamicServiceRegistrySupplier");
+            boolean isDroolsStaticPresent = isClassDefined("org.drools.statics.StaticServiceRegistrySupplier");
```

Comment:
Nitpick: Personally, I'd inline these booleans in the if statements, as they are only used once.

Reply:
Myself, I don't see the DRY principle as the only motivation for creating methods or, in this case, local variables. The variable name clearly states what's the purpose of the check. In contrast to that, `isClassDefined("org.drools.dynamic.DynamicServiceRegistrySupplier")` is not so clear about it's purpose as I just deliberately picked the class `DynamicServiceRegistrySupplier` as something pointing to the presence of the `drools-core-dynamic`. The local variables serve here rather in the role of comments.